File: sdk/php/examples/contract_smoke.php
<?php

declare(strict_types=1);

/**
 * Contract smoke for the PHP SDK — exercises the public API against a live
 * backend and prints exactly one JSON line on stdout.
 *
 * Run from the repository root (no composer install needed):
 *   EXPERIMENTLY_API_KEY=<key> php sdk/php/examples/contract_smoke.php
 * or through the live runner, which seeds the fixtures and supplies the key:
 *   python tests/sdk-contract/live/run_live_contract.py --sdk php --strict
 *
 * Environment:
 *   EXPERIMENTLY_API_URL     backend origin (default http://localhost:8000)
 *   EXPERIMENTLY_API_KEY     API key (required)
 *   CONTRACT_EXPERIMENT_KEY  default sdk_contract_ab
 *   CONTRACT_FLAG_KEY        default sdk_contract_flag
 *   CONTRACT_USER_ID         default smoke-<uuid>
 *
 * On success: one JSON line on stdout, exit 0. On failure: one line on stderr,
 * exit 1. Nothing else is ever written to stdout (PHP notices go to stderr).
 */

use ExperimentationPlatform\ExperimentationClient;
use ExperimentationPlatform\SdkConfig;

// Keep stdout clean: any PHP warning/notice/deprecation goes to stderr.
error_reporting(E_ALL);
ini_set('display_errors', 'stderr');

$autoload = __DIR__ . '/../vendor/autoload.php';
if (is_file($autoload)) {
    require_once $autoload;
} else {
    // Dependency order matters without an autoloader: base classes first.
    foreach (
        [
            'Errors/ExperimentationException.php',
            'Errors/ApiException.php',
            'Errors/AuthException.php',
            'Errors/NetworkException.php',
            'SdkConfig.php',
            'HttpClient.php',
            'Cache.php',
            'FeatureFlagEvaluator.php',
            'FlagEvaluation.php',
            'Assignment.php',
            'BatchResult.php',
            'ExperimentationClient.php',
        ] as $file
    ) {
        require_once __DIR__ . '/../src/' . $file;
    }
}

function smokeFail(string $message): never
{
    fwrite(STDERR, "contract_smoke(php): {$message}\n");
    exit(1);
}

function envOr(string $name, string $default): string
{
    $value = getenv($name);
    return ($value === false || $value === '') ? $default : $value;
}

/** Random RFC 4122 v4 UUID prefixed with "smoke-". */
function randomUserId(): string
{
    $bytes = random_bytes(16);
    $bytes[6] = chr((ord($bytes[6]) & 0x0f) | 0x40);
    $bytes[8] = chr((ord($bytes[8]) & 0x3f) | 0x80);
    return 'smoke-' . vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($bytes), 4));
}

/**
 * Runs the four contract steps through the SDK's public API and returns the report.
 *
 * @return array<string, mixed>
 */
function runSmoke(string $apiUrl, string $apiKey, string $experimentKey, string $flagKey, string $userId): array
{
    $client     = new ExperimentationClient(new SdkConfig(baseUrl: $apiUrl, apiKey: $apiKey, timeout: 10));
    $attributes = ['source' => 'contract_smoke', 'sdk' => 'php'];

    // 1. Sticky assignment: assign twice, dropping the local cache in between so
    //    the second answer really comes from the server.
    $first = $client->getAssignment($experimentKey, $userId, $attributes);
    if ($first === null) {
        smokeFail("assignment failed for {$experimentKey} (is the experiment ACTIVE and the API key valid?)");
    }
    $client->clearCache();
    $second = $client->getAssignment($experimentKey, $userId, $attributes);
    if ($second === null) {
        smokeFail("second assignment failed for {$experimentKey}");
    }
    if (!in_array($first->variantName, ['control', 'treatment'], true)) {
        smokeFail('unexpected variant_name ' . var_export($first->variantName, true));
    }
    $sticky = $first->variantName === $second->variantName && $first->variantId === $second->variantId;
    if (!$sticky) {
        smokeFail("assignment not sticky: {$first->variantName} then {$second->variantName}");
    }

    // 2. Flag evaluation (the seeded flag is 100% on).
    $flag = $client->evaluateFlag($flagKey, $userId);
    if ($flag->enabled !== true) {
        smokeFail("flag {$flagKey} evaluated as " . var_export($flag->enabled, true) . ', expected true');
    }

    // 3. Track with an experiment key -> one POST /api/v1/tracking/track.
    if (!$client->track('purchase', $userId, ['source' => 'contract_smoke'], $experimentKey, null, 12.5)) {
        smokeFail('track(purchase) with experiment_key failed');
    }

    // 4. Track without a key -> POST /api/v1/tracking/batch fanned out to the
    //    cached assignment + evaluated flag.
    if (!$client->track('page_view', $userId, ['page' => '/smoke'])) {
        smokeFail('track(page_view) fan-out failed');
    }

    // 4b. Explicit two-event batch.
    $batch = $client->trackBatch([
        ['event_name' => 'page_view', 'user_id' => $userId, 'experiment_key' => $experimentKey, 'properties' => ['page' => '/batch']],
        ['event_name' => 'page_view', 'user_id' => $userId, 'feature_flag_key' => $flagKey, 'properties' => ['page' => '/batch']],
    ]);
    if (!$batch->isOk()) {
        smokeFail('trackBatch failed: ' . json_encode($batch->errors));
    }

    return [
        'sdk'    => 'php',
        'assign' => ['variant_name' => $first->variantName, 'is_control' => $first->isControl, 'sticky' => $sticky],
        'flag'   => ['enabled' => $flag->enabled],
        'track'  => ['ok' => true],
        'fanout' => ['ok' => true],
    ];
}

$apiKey = envOr('EXPERIMENTLY_API_KEY', '');
if (trim($apiKey) === '') {
    smokeFail('EXPERIMENTLY_API_KEY is required');
}

try {
    $report = runSmoke(
        envOr('EXPERIMENTLY_API_URL', 'http://localhost:8000'),
        $apiKey,
        envOr('CONTRACT_EXPERIMENT_KEY', 'sdk_contract_ab'),
        envOr('CONTRACT_FLAG_KEY', 'sdk_contract_flag'),
        envOr('CONTRACT_USER_ID', randomUserId()),
    );
    echo json_encode($report, JSON_THROW_ON_ERROR), "\n";
} catch (Throwable $e) {
    smokeFail($e::class . ': ' . $e->getMessage());
}

exit(0);

File: sdk/ruby/examples/contract_smoke.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Contract smoke for the Ruby SDK — exercises the public API against a live
# backend and prints exactly one JSON line on stdout.
#
# Run from the repository root:
#   EXPERIMENTLY_API_KEY=$(cat tests/sdk-contract/live/.api_key) \
#     ruby -Isdk/ruby/lib sdk/ruby/examples/contract_smoke.rb
#
# Environment:
#   EXPERIMENTLY_API_URL     backend origin (default http://localhost:8000)
#   EXPERIMENTLY_API_KEY     API key (required)
#   CONTRACT_EXPERIMENT_KEY  default sdk_contract_ab
#   CONTRACT_FLAG_KEY        default sdk_contract_flag
#   CONTRACT_USER_ID         default smoke-<uuid>

require 'json'
require 'securerandom'
require 'experimently'

def smoke_fail(message)
  $stderr.puts "contract_smoke(ruby): #{message}"
  exit 1
end

api_url        = ENV.fetch('EXPERIMENTLY_API_URL', 'http://localhost:8000')
api_key        = ENV['EXPERIMENTLY_API_KEY']
experiment_key = ENV.fetch('CONTRACT_EXPERIMENT_KEY', 'sdk_contract_ab')
flag_key       = ENV.fetch('CONTRACT_FLAG_KEY', 'sdk_contract_flag')
user_id        = ENV['CONTRACT_USER_ID'] || "smoke-#{SecureRandom.uuid}"

smoke_fail('EXPERIMENTLY_API_KEY is required') if api_key.nil? || api_key.strip.empty?

client = Experimently::Client.new(base_url: api_url, api_key: api_key, timeout: 10)

# 1. Sticky assignment: assign twice, dropping the local cache in between so the
#    second answer really comes from the server.
first = client.get_assignment(experiment_key, user_id, { source: 'contract_smoke' })
smoke_fail("assignment failed for #{experiment_key} (is the experiment ACTIVE and the key valid?)") if first.nil?
client.clear_cache
second = client.get_assignment(experiment_key, user_id, { source: 'contract_smoke' })
smoke_fail("second assignment failed for #{experiment_key}") if second.nil?

unless %w[control treatment].include?(first.variant_name)
  smoke_fail("unexpected variant_name #{first.variant_name.inspect}")
end
sticky = first.variant_name == second.variant_name && first.variant_id == second.variant_id
smoke_fail("assignment not sticky: #{first.variant_name} then #{second.variant_name}") unless sticky

# 2. Flag evaluation (the seeded flag is 100% on).
flag = client.evaluate_flag(flag_key, user_id)
smoke_fail("flag #{flag_key} evaluated as #{flag.enabled.inspect}, expected true") unless flag.enabled == true

# 3. Track with an experiment key -> one /tracking/track.
track_ok = client.track('purchase', user_id,
                        properties: { source: 'contract_smoke' },
                        experiment_key: experiment_key,
                        value: 12.5)
smoke_fail('track(purchase) failed') unless track_ok

# 4. Track without a key -> /tracking/batch fan-out to the cached assignment + flag.
fanout_ok = client.track('page_view', user_id, properties: { page: '/smoke' })
smoke_fail('track(page_view) fan-out failed') unless fanout_ok

# 4b. Explicit two-event batch.
batch = client.track_batch([
  { event_name: 'page_view', user_id: user_id, experiment_key: experiment_key, properties: { page: '/batch' } },
  { event_name: 'page_view', user_id: user_id, feature_flag_key: flag_key, properties: { page: '/batch' } }
])
smoke_fail("track_batch failed: #{batch.errors.inspect}") unless batch.ok?

puts JSON.generate(
  sdk:    'ruby',
  assign: { variant_name: first.variant_name, is_control: first.is_control, sticky: true },
  flag:   { enabled: flag.enabled },
  track:  { ok: true },
  fanout: { ok: true }
)
exit 0
